navigation/navigation-compose/src/main/java/androidx/navigation/compose/NavBackStackEntryProvider.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.navigation.compose
+
+import android.annotation.SuppressLint
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.runtime.saveable.SaveableStateHolder
+import androidx.compose.ui.platform.LocalLifecycleOwner
+import androidx.compose.ui.platform.LocalSavedStateRegistryOwner
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewmodel.compose.LocalViewModelStoreOwner
+import androidx.lifecycle.viewmodel.compose.viewModel
+import androidx.navigation.NavBackStackEntry
+import java.util.UUID
+
+
+/**
+ * Provides [this] [NavBackStackEntry] as [LocalViewModelStoreOwner], [LocalLifecycleOwner] and
+ * [LocalSavedStateRegistryOwner] to the [content] and saves the [content]'s saveable states with
+ * the given [saveableStateHolder].
+ *
+ * @param saveableStateHolder The [SaveableStateHolder] that holds the saved states
+ * @param content The content [Composable]
+ */
+@SuppressLint("ComposableNaming")
+@Composable
+public fun NavBackStackEntry.provideToCompositionLocals(
+    saveableStateHolder: SaveableStateHolder,
```

Comment:
Would this work for the simple case if it had a default value of `rememberSaveableStateHolder()`?

Reply:
Yeah - I chose not to provide a default value because I felt it might discourage hoisting the `SaveableStateHolder` but I guess that's alright!

Reply:
I am actually also not sure have a default value here is reasonable. the whole SaveableStateHolder logic is pretty much useless if it is not hoisted somewhere before iterating through all the entries

Reply:
I was thinking about the case in #173 where only the latest destination is shown (multiple bottom sheets at the same time doesn't make much sense) where having this default would make this more self contained.

I'd be fine leaving out the default and having developers have to think about the hoisting of their `SaveableStateHolder` if you think that better.

Reply:
We could leave out the default param and specify in the KDocs that the `saveableStateHolder` should be hoisted? Without that remark folks probably wouldn't actually hoist this.

Reply:
the same SaveableStateHolder should be used for all the backstack entries. with code like this
backStack.forEach {
    it.CompositionLocalProvider(rememberSaveableStateHolder()) {...}
}
we have different holder for each backstack, which is not how it is expected to be used.
so I would vote to remove the default param value just to not add extra confusion (we can explain it with more details in the doc as well)

Reply:
@andkulikov Decided to go with `LocalOwnersProvider` as it did make more sense after thinking about it a bit more :) If anybody has any other ideas we can still rename

Also made the param non-optional and improved docs in a8100a0